src/ripple/app/misc/impl/ValidatorList.cpp

```diff
@@ -17,174 +17,744 @@
+        std::lock_guard<std::mutex> lock{mutex_};
+        stopping_ = true;
     }
+    if(auto sp = work_.lock())
+        sp->cancel();
+    error_code ec;
+    timer_.cancel(ec);
+    stopping_ = false;
+}
 
-    return ephemeral_.emplace (identity, comment).second;
+bool
+ValidatorList::listed (
+    PublicKey const& identity) const
+{
+    auto const masterKey = manifests_.getMasterKey (identity);
+    auto const pubKey = masterKey ? *masterKey : identity;
+    return keyListings_.find (pubKey) != keyListings_.end ();
 }
 
 bool
-ValidatorList::removeEphemeralKey (
-    PublicKey const& identity)
+ValidatorList::trusted (PublicKey const& identity) const
 {
-    std::lock_guard <std::mutex> sl (mutex_);
-    return ephemeral_.erase (identity);
+    auto const signingKey = manifests_.getSigningKey (identity);
+    auto const pubKey = signingKey ? *signingKey : identity;
+    return trustedKeys_.find (pubKey) != trustedKeys_.end();
 }
 
 bool
-ValidatorList::insertPermanentKey (
-    PublicKey const& identity,
-    std::string const& comment)
+ValidatorList::trustedPublisher (PublicKey const& identity) const
 {
-    std::lock_guard <std::mutex> sl (mutex_);
+    return publisherLists_.count (identity);
+}
 
-    if (ephemeral_.find (identity) != ephemeral_.end())
-    {
-        JLOG (j_.error()) <<
-            toBase58 (TokenType::TOKEN_NODE_PUBLIC, identity) <<
-            ": permanent key exists in ephemeral table!";
+std::size_t
+ValidatorList::calcQuorum (
+    std::uint32_t nTrusted,
+    std::uint32_t nListed)
+{
+    const std::vector<std::size_t> neededValidations{
+        0, 1, 2, 3, 3, 4, 5, 5, 6, 7, 7};
+    auto constexpr quorumRatio = 0.8;
+
+    // Floor of 80% of minimum required overlap with all listed validators
+    auto quorum = std::ceil (
+        quorumRatio *
+        std::ceil (
+            MINIMUM_NETWORK_OVERLAP *
+            std::max (nListed, nTrusted)));
+
+    // If >10 trusted validation public keys, set quorum to 80% of the number
+    // of trusted keys.
+    // Otherwise, use lookup table with quorums for byzantine fault tolerance
+    // (3f+1)
+    std::size_t nVals = (nTrusted > 10) ?
+        std::ceil (nTrusted * quorumRatio) :
+        neededValidations[nTrusted];
+
+    return (quorum < nVals) ? nVals : quorum;
+}
+
+bool
+ValidatorList::removePublisher (PublicKey const& publisherKey)
+{
+    std::lock_guard <std::mutex> lock{mutex_};
+
+    auto const iList = publisherLists_.find (publisherKey);
+    if (iList == publisherLists_.end ())
         return false;
+
+    JLOG (j_.debug()) <<
+        "Removing validator list for revoked publisher " <<
+        toBase58(TokenType::TOKEN_NODE_PUBLIC, publisherKey);
+
+    for (auto const& val : iList->second.list)
+    {
+        auto const& iVal = keyListings_.find (val);
+        if (iVal == keyListings_.end())
+            continue;
+
+        if (iVal->second <= 1)
+            keyListings_.erase (iVal);
+        else
+            --iVal->second;
```

Comment:
I find this particular statement hard to read - something like `iVal->second -= 1` makes more sense to me, but I don't feel strongly about it.